File: LRUCache.cpp
//http://www.cs.uml.edu/~jlu1/doc/codes/lruCache.html
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;
struct LRUCacheEntry
{
	int key;
	int data;
	LRUCacheEntry* prev;
	LRUCacheEntry* next;
};

class LRUCache
{
private:
	unordered_map< int, LRUCacheEntry* >	_mapping;
	vector< LRUCacheEntry* >		_freeEntries;
	LRUCacheEntry *			head;
	LRUCacheEntry *			tail;
	LRUCacheEntry *			entries;
public:
	LRUCache(size_t size){
		entries = new LRUCacheEntry[size];
		for (int i=0; i<size; i++)
			_freeEntries.push_back(entries+i);
		head = new LRUCacheEntry;
		tail = new LRUCacheEntry;
		head->prev = NULL;
		head->next = tail;
		tail->next = NULL;
		tail->prev = head;
	}
	~LRUCache()
	{
		delete head;
		delete tail;
		delete [] entries;
	}
	void set(int key, int data)
	{
		LRUCacheEntry* node = _mapping[key];
		if(node)
		{
			// refresh the link list
			detach(node);
			node->data = data;
			attach(node);
		}
		else{
			if ( _freeEntries.empty() )
			{
				node = tail->prev;
				detach(node);
				_mapping.erase(node->key);
				node->data = data;
				node->key = key;
				_mapping[key] = node;
				attach(node);
			}
			else{
				node = _freeEntries.back();
				_freeEntries.pop_back();
				node->key = key;
				node->data = data;
				_mapping[key] = node;
				attach(node);
			}
		}
	}

	int get(int key)
	{
		LRUCacheEntry* node = _mapping[key];
		if(node)
		{
			detach(node);
			attach(node);
			return node->data;
		}
		else return -1;
	}

private:
	void detach(LRUCacheEntry* node)
	{
		node->prev->next = node->next;
		node->next->prev = node->prev;
	}
	void attach(LRUCacheEntry* node)
	{
		node->next = head->next;
		node->prev = head;
		head->next = node;
		node->next->prev = node;
	}
};

File: LRUCache.java
class LRUCacheEntry{
    int key ;
    int value;
    LRUCacheEntry next = null;
    LRUCacheEntry prev = null;
}
public class LRUCache {
    Map<Integer,LRUCacheEntry> _mapping;
    List<LRUCacheEntry> _freeEntries; 
    LRUCacheEntry head,tail;
    LRUCacheEntry [] entries ;
    
    
    public LRUCache(int capacity) {
         _mapping = new HashMap<>();

        _freeEntries  = new ArrayList<>();
        entries = new LRUCacheEntry[capacity];
        for(int i =0;i<capacity;++i){
            entries[i] = new LRUCacheEntry();
            _freeEntries.add(entries[i]);
        }
        head = new LRUCacheEntry();
        tail = new LRUCacheEntry();
        head.next = tail;
        head.prev = null;
        tail.prev = head;
        tail.next = null;
    }
    
    public int get(int key) {
        LRUCacheEntry node = _mapping.get(key);
        if(node!=null){
            detachNode(node);
            attachNode(node);
            return node.value;
        }
        return -1;
        
    }
    
    public void put(int key, int value) {
        LRUCacheEntry node = _mapping.get(key);
        if(node!=null){
            detachNode(node);
            node.value = value;
            attachNode(node);
            return;
        }
        if(_freeEntries.isEmpty()){
            node = tail.prev;
            _mapping.remove(node.key);
            node.key = key;
            node.value = value;
            detachNode(node);
            attachNode(node);
            _mapping.put(key,node);

        }
        else{
            
             node = _freeEntries.get(_freeEntries.size()- 1);
            _freeEntries.remove(_freeEntries.size() - 1);
            node.key = key;
            node.value = value;
           
            _mapping.put(key,node);
            attachNode(node);
        }
    }
    public void detachNode(LRUCacheEntry node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
    public void attachNode(LRUCacheEntry node){
        node.next = head.next;
        head.next.prev = node;
        head.next = node;
        node.prev = head;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
